refactor acceptance/version-upgrade

tbg opened the pull request:
Please refer to the individual commit messages.
In summary, this test is now polished, avoids starting ancient binaries while
retaining the data set created by ancient binaries, and can be used to cheaply
create other mixed-version roachtests.

Release note: None

tbg committed:
roachtest: refactor acceptance/version-upgrade

This test has not aged well. As a first step, clean it up by introducing
explicit state for this test and rearranging the code so that readers
get an idea of how it works more easily.

The next step will be to refactor the feature tests to be more general;
as it stands, they're pretty useless because in a feature test you don't
know what version you're running against and so it's hard to reason
about what to expect.

Release note: None

tbg committed:
roachtest: create fixtures for acceptance/version-upgrade

This will allow us to stop running old versions in this test while at
the same time continuing to test clusters that originated on ancient
versions of CRDB.

Release note: None

tbg committed:
roachprod: expand destination in put

Allow `roachprod put source {store-dir}/dest`.

Release note: None

tbg committed:
roachtest: bake <19.1 into acceptance/version-upgrade

This test used to start a v1.0 cluster and migrated it all the way
to the the master branch. As a result, the test would take longer
to run with each release, but additionally it would sometimes flake
in ancient releases that we don't support any more.

Instead of just ditching the old releases altogether, we now start
the test up from store directories that were initiated at v1.0 and
were upgraded all the way through the lowest version that we care
to run in this test. This is different from bootstrapping a new
cluster at that version since older versions may have left old
data around, or never gotten rid of idiosyncracies that we have since
fixed.

Touches #44732.

Release note: None

tbg committed:
roachtest: remove workarounds from acceptance/version-upgrade

Now that this test does not run ancient CRDB binaries any more, we get
to delete a number of workarounds.

Release note: None

tbg committed:
roachtest: refactor feature tests in acceptance/version-upgrade

Switch to a more powerful notion of feature flags that has the freedom
to do "basically anything".

Release note: None

tbg committed:
roachtest: simplify init in version-upgrade

Avoid hard-coding the fixture load in the test helper. My hope is that
the test helper can be used to cheaply set up dedicated mixed version
tests.

Also perform some renames and a round of commentary to explain better
how feature tests work.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46924)
<!-- Reviewable:end -->

tbg committed:
roachtest: indicate when mixed feature test skips itself

Forgot to hook this up earlier. At the time of writing, nothing is
skipped because we're not running these old versions any more.

Release note: None

andreimatei reviewed (approved):
LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46924#-:-M3vOFk69ZB3ZnOmzZK4:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 1055 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46924#-M3vLO6yETNBOc2932_y:-M3vLO6yETNBOc2932_z:b-8m3ry7) ([raw file](https://github.com/cockroachdb/cockroach/blob/152114fb56a704b72c12b48688ec54f13f0a1c35/pkg/cmd/roachprod/install/cluster_synced.go#L1055)):*
> ```Go
> 					return
> 				}
> 				// TOOD(jlinder): this does not take into account things like
> ```

hmmm how about taking this opportunity to mimic `cp` behavior? :)
___
*[pkg/cmd/roachtest/upgrade.go, line 573 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46924#-M3vMaZq6i_Pd8XUT4hX:-M3vMaZq6i_Pd8XUT4hY:bxs5bac) ([raw file](https://github.com/cockroachdb/cockroach/blob/cde0d909a32d403dac4573cad00cca871eda630b/pkg/cmd/roachtest/upgrade.go#L573)):*
> ```Go
> }
> 
> // waitForUpgradeStep waits for the cluster version to reach that reflected by the first node. We
> ```

nit: wrap
___
*[pkg/cmd/roachtest/upgrade.go, line 573 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46924#-M3vMfwZ34y8PaxqeJA_:-M3vMfwZ34y8PaxqeJAa:b5ipdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/cde0d909a32d403dac4573cad00cca871eda630b/pkg/cmd/roachtest/upgrade.go#L573)):*
> ```Go
> }
> 
> // waitForUpgradeStep waits for the cluster version to reach that reflected by the first node. We
> ```

word missing in first sentence?
___
*[pkg/cmd/roachtest/upgrade.go, line 578 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46924#-M3vNa0d5Aj-yXw6ZMxl:-M3vNa0d5Aj-yXw6ZMxm:bw4ukzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cde0d909a32d403dac4573cad00cca871eda630b/pkg/cmd/roachtest/upgrade.go#L578)):*
> ```Go
> // some nodes may already have learned of an upgrade (via Gossip) while others have not. This
> // situation tends to exhibit unexpected behavior.
> func waitForUpgradeStep() versionStep {
> ```

Can you spell out in the comment that by "separate step" you mean separating this from the previous step which bumps the version?

And can we actually do something to block gossip, to ensure that we actually run the feature tests while some nodes are not aware of the version bump? Cause otherwise, are we really running in that mode?


<!-- Sent from Reviewable.io -->

tbg commented:
TFTR!
I won't wrangle roachprod put today. There's no testing and the opportunity cost is too high.

We are exposing tests to the mixed gossip state mode only heuristically. It's an interesting idea to delay the gossip update for the version bump, I will think about that while I fix the schema change flake in this test.

bors r=andreimatei
